### PR TITLE
Feature: MFA for admins

### PR DESCRIPTION
**Pull request by sfnelson**

### CODE REVIEW

**Review by mattr — APPROVED** [review]
LGTM. Nice to see this flow enabled.

Some minor thoughts for additional tests, possible security, and clarity.

File: app/views/admin/admin_users/show.html+self.erb [hunk 1]
```diff
@@ -4,11 +4,25 @@
...
+    <span class="repel">
+      <%= otp %>
+      <% if otp.value %>
+        <%= button_to("Remove", admin_admin_user_otp_path(admin),
+                      class:  "button button--text",
+                      method: :delete,
+                      form:   { data: { turbo_confirm: "Are you sure?" } }) %>
+      <% else %>
+        <%= kpop_link_to "Add", new_admin_admin_user_otp_path(admin) %>
+      <% end %>
```
[the remark this answers is not in the document]
Comment (sfnelson) [reply app/views/admin/admin_users/show.html+self.erb:23]: @mattr this template has the `+self` variant

File: app/models/concerns/koi/model/otp.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+module Koi
+  module Model
+    module OTP
+      extend ActiveSupport::Concern
+
+      def requires_otp?
+        otp_secret.present?
+      end
```
[the remark this answers is not in the document]
Comment (sfnelson) [reply app/models/concerns/koi/model/otp.rb:14]: Yes I'd like to, but I don't want to require that Koi applications enable row-level encryption just for this feature. The secret is only relevant to this application so the risk if someone was to steal a database dump is low (only impacts this Koi install).

File: app/controllers/admin/sessions_controller.rb [hunk 3]
```diff
@@ -14,7 +14,46 @@ def new
...
+
+      session[:admin_user_id] = nil
+
+      redirect_to new_admin_session_path
+    end
+
+    private
+
+    def create_session_with_password
+      return if session_params[:password].blank?
```
[the remark this answers is not in the document]
Comment (sfnelson) [reply app/controllers/admin/sessions_controller.rb:47]: It's a hangover from the previous code, removed

File: app/views/admin/admin_users/show.html+self.erb (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattr) [reply app/views/admin/admin_users/show.html+self.erb:23]: I mean, that would also work. 🤦

**Pull request merged**